src/core/memory.d

```diff
@@ -895,6 +895,8 @@ void pureFree(void* ptr) @system pure @nogc nothrow
     // See also: https://issues.dlang.org/show_bug.cgi?id=17956
     void* z = pureMalloc(size_t.max & ~255); // won't affect `errno`
     assert(errno == fakePureErrno()); // errno shouldn't change
+
+    version (DragonFlyBSD) {} else    // FIXME : z seems to be allocated and not null. Why would size_t.max - 2 not be allowed to allocate ? or is the assertion wrong
     assert(z is null);
```

Comment:
z seems to be allocated and not null.
Why would size_t.max - 2 not be allowed to allocate ? or is the assertion wrong

Reply:
I can't find any documented reference, but I believe this is testing if the glibc's malloc would fail because of the internal block size and the additional alignment (and any other internal metadata) that should be added to the internal allocation, which while adding with the values close to `size_t.max` would lead to the overflow.

Reply:
Wrote a little test program in ansi-c
```c
#include <stdio.h>
#include <stdlib.h>

// On Linux
#if !defined(SIZE_MAX)
#define SIZE_MAX __SIZE_MAX__		// ((size_t)-1) = 9223372036854775807 = LONG_MAX
#endif
// On DragonFlyBSD
//#define SIZE_MAX UINT64_MAX  // from /usr/include/machine/int_limits.h

int main () {
  printf("SIZE_MAX=%lu\n", SIZE_MAX);
  
  unsigned long tests[] = {
    9223372036854775807,
    223372036854775808,
    23372036854775808,
    3372036854775808,
    372036854775808,
    72036854775808,
    2036854775808,
    36854775808,
    SIZE_MAX - 2,                                    // The previous value for the test // https://issues.dlang.org/show_bug.cgi?id=17956
    SIZE_MAX,
    SIZE_MAX + 'a',                                  // simulating SIZE_MAX ~ 255
    SIZE_MAX + SIZE_MAX,
    SIZE_MAX * 20};
  
  for (int test = 0; test < 12; test++) {
    printf("test%d: malloc size:%lu\n", test, tests[test]);

    int *ptr = (int *) malloc(tests[test]);
    if (ptr == NULL) {
        printf("malloc failed, ptr == NULL\n");
    } else {
        printf(" ptr == %p\n", ptr);
        free(ptr);
        ptr = NULL;
    }
  }
  return 0;
}
```
Compiled using `gcc test_malloc.c -o test_malloc`

Linux x86_64 Results:
```
SIZE_MAX=18446744073709551615
test0: malloc size:9223372036854775807
malloc failed, ptr == NULL
test1: malloc size:223372036854775808
malloc failed, ptr == NULL
test2: malloc size:23372036854775808
malloc failed, ptr == NULL
test3: malloc size:3372036854775808
malloc failed, ptr == NULL
test4: malloc size:372036854775808
malloc failed, ptr == NULL
test5: malloc size:72036854775808
malloc failed, ptr == NULL
test6: malloc size:2036854775808
malloc failed, ptr == NULL
test7: malloc size:36854775808
malloc failed, ptr == NULL
test8: malloc size:18446744073709551613
malloc failed, ptr == NULL
test9: malloc size:18446744073709551615
malloc failed, ptr == NULL
test10: malloc size:96
 ptr == 0x7fd0700008c0
test11: malloc size:18446744073709551614
malloc failed, ptr == NULL
```

DragonFlyBSD (x86_64) Resuls:
```
SIZE_MAX=18446744073709551615
test0: malloc size:9223372036854775807
malloc failed, ptr == NULL
test1: malloc size:223372036854775808
malloc failed, ptr == NULL
test2: malloc size:23372036854775808
malloc failed, ptr == NULL
test3: malloc size:3372036854775808
malloc failed, ptr == NULL
test4: malloc size:372036854775808
malloc failed, ptr == NULL
test5: malloc size:72036854775808
 ptr == 0x800800000
test6: malloc size:2036854775808
 ptr == 0x800800000
test7: malloc size:36854775808
 ptr == 0x800800000
test8: malloc size:18446744073709551613
 ptr == 0x800455000
test9: malloc size:18446744073709551615
 ptr == 0x800455000
test10: malloc size:96
 ptr == 0x8006200c0
test11: malloc size:18446744073709551614
 ptr == 0x800455000
```
On DragonFly it seems to be ok, as long as the upperbound of '9223372036854775807' is observed. Using SIZE_MAX however results faulty behaviour.

Reply:
Perhaps is worth raising the bug report in the DragoFlyBSD tracker?

Reply:
After creating the above 'proof' i think that makes sense.
Once the bug report i made, i can attach a link to the 'unittest exclusion', so it can be tracked.

Reply:
Found a potential solution to the malloc(SIZE_MAX) issue. Send a diff to upstream dragonfly issue manager (Issue [3114](https://bugs.dragonflybsd.org/issues/3114)), seems to resolve the case reported above.

Reply:
Issue has been fixed in dragonfly master
Removing this unittest escape